pkg/process/util/containers.go

```diff
@@ -6,104 +6,280 @@
 package util
 
 import (
+	"sync"
 	"time"
 
-	"github.com/DataDog/datadog-agent/pkg/util/cache"
+	model "github.com/DataDog/agent-payload/v5/process"
+
+	"github.com/DataDog/datadog-agent/pkg/tagger"
+	"github.com/DataDog/datadog-agent/pkg/tagger/collectors"
 	"github.com/DataDog/datadog-agent/pkg/util/containers"
-	"github.com/DataDog/datadog-agent/pkg/util/containers/collectors"
+	"github.com/DataDog/datadog-agent/pkg/util/containers/v2/metrics"
 	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-agent/pkg/util/system"
+	"github.com/DataDog/datadog-agent/pkg/workloadmeta"
+)
+
+const (
+	floatNanoseconds float64 = float64(time.Second)
 )
 
+// ContainerRateMetrics holds previous values for a container,
+// in order to compute rates
+type ContainerRateMetrics struct {
+	UserCPU            float64
+	SystemCPU          float64
+	TotalCPU           float64
+	IOReadBytes        float64
+	IOWriteBytes       float64
+	NetworkRcvdBytes   float64
+	NetworkSentBytes   float64
+	NetworkRcvdPackets float64
+	NetworkSentPackets float64
+}
+
+// NullContainerRates can be safely used for containers that have no
+// previous rate values stored (new containers)
+var NullContainerRates = ContainerRateMetrics{
+	UserCPU:   -1,
+	SystemCPU: -1,
+	TotalCPU:  -1,
+}
+
+// ContainerProvider provides data about containers usable by process-agent
+type ContainerProvider struct {
```

Comment:
What do you think about changing this to be an interface that can be backed by this implementation but also by a mock, currently we don't seem to have a good way to test the container checks without scaffolding container environment.

What I am proposing in the nutshell:
```
type ContainerProvider interface {
	GetContainers(cacheValidity time.Duration, previousContainers map[string]*ContainerRateMetrics, previousTime, currentTime time.Time) ([]*model.Container, map[string]*ContainerRateMetrics, map[int]string, error)
}

type containerProvider struct {
	metricsProvider metrics.Provider
	metadataStore   workloadmeta.Store
	filter          *containers.Filter
}

Reply:
Good idea, implemented.